### ISSUE TITLE
[RC2] enable crossgen of mscorlib.dll on Linux and Mac

### ISSUE THREAD

**Issue by joshfree** [dotnet/coreclr#1934]
Update the xplat builds to produce official mscorlib.ni.dll binaries for Linux and mac

/cc @chcosta @gkhanna79

**Comment by gkhanna79** [comment]
This is blocked against MSbuild support being available.

**Comment by gkhanna79** [comment]
@weshaggard Do we have MSbuild working on coreCLR for all platforms?

**Comment by weshaggard** [comment]
For CoreFX it is being tracked by https://github.com/dotnet/corefx/issues/5021. I think Msbuild on .NET Core mostly works but there isn't anyone working on integrating that into the coreclr builds for mscorlib. My suspicion is that this will not be done in the RC2 timeframe. 

cc @joperezr

**Comment by gkhanna79** [comment]
If Msbuild has been confirmed to be successfully integrated for CoreFX builds and known to function properly, @JohnChen0 will work on CoreCLR repo using it. However, it is not understood what is the current state of affairs - last schedule was it being available post Thanksgiving.

> My suspicion is that this will not be done in the RC2 timeframe. 

Are you saying MSbuild support will come in post RC2?

**Comment by weshaggard** [comment]
I was referring to changes needed for converting the coreclr repo. If @JohnChen0 can do that work then we should be able to get this working. @JohnChen0 please work with @joperezr on how to get a version of msbuild that runs on .NET Core. It is my understanding that is in pretty good working state.

**Comment by gkhanna79** [comment]
Awesome! @JohnChen0 will look into this.

**Comment by JohnChen0** [comment]
I was able to build mscorlib.dll on Linux using work-in-progress code from Jose. There are still some known issues and work items:
* I used private NuGet packages created by Daniel.
* MSBuild attempts to call a Windows registry API function during cleanup. This happens after mscorlib.dll has already been created, but it causes MSBuild to crash.This is tracked by issue Microsoft/msbuild#423.
* Csc task in msbuild attempts to use corerun.exe instead of corerun on Linux.
* Various steps still need to be automated.
* Make dotnet CLI a prerequisite of building CoreCLR, since we use “dotnet restore” and “dotnet publish” to install MSBuild.
* Once the build of mscorlib.dll is fully working, need to add the step to crossgen it.

**Comment by gkhanna79** [comment]
> I used private NuGet packages created by Daniel.

@weshaggard will be helping with this.

> MSBuild attempts to call a Windows registry API function during cleanup. This happens after mscorlib.dll has already been created, but it causes MSBuild to crash.This is tracked by issue Microsoft/msbuild#423.

> Csc task in msbuild attempts to use corerun.exe instead of corerun on Linux.

@weshaggard do you know who is working on fixing CSC?

**Comment by gkhanna79** [comment]
> Various steps still need to be automated.
@JohnChen0 Can you please share more details on this?

> Make dotnet CLI a prerequisite of building CoreCLR, since we use “dotnet restore” and “dotnet publish” to install MSBuild.
@JohnChen0 Can we have change for this ready?

> Once the build of mscorlib.dll is fully working, need to add the step to crossgen it.
Lets get the change ready for this as well.

**Comment by JohnChen0** [comment]
The latest WIP code is at 6f98b80. @joperezr, @gkhanna79, @dsplaisted, @rainersigwald: Could you please take a look? The code works, but with some known issues:
* Still using a private NuGet package for Microsoft.Net.Compilers.Targets.NetCore. @dsplaisted: when will an equivalent package be available on a feed?
* Currently using a local copy of init-msbuild.sh. Eventually this should come from the BuildTools package.
* MSBuild is still affected by Microsoft/msbuild#423, thus I have to ignore the return status from MSBuild. @rainersigwald: any update on when a fix will be available?
* The change has been tested on Linux, but I don't have a Mac to test it on. Could anyone help with testing it on Mac?
* I haven't been able to get MSBuild to build mscorlib on Windows. I think having this working on Windows should be considered a nice-to-have feature.

**Comment by rainersigwald** [comment]
@JohnChen0 I just pushed a fix for https://github.com/Microsoft/msbuild/issues/423. @AndyGerlicher is looking at fixing builds and we'll get an updated package published ASAP.

**Comment by JohnChen0** [comment]
Created issue #2698 for getting a package for Microsoft.Net.Compilers.Targets.NetCore. Setting up MSBuild to build mscorlib on Linux is blocked by #2698.

**Comment by JohnChen0** [comment]
Update WIP code to new commit 94ff2e4, using the latest NuGet package for MSBuild. Verified that issue Microsoft/msbuild#423 is fixed.

**Comment by JohnChen0** [comment]
Two MSBuild issues are blocking resolving this issue on Mac OS X:
* MSBuild does not accept absolute path on command line on Mac (Microsoft/msbuild#445)
* The OS property is different between CoreCLR version and mono version of MSBuild (Microsoft/msbuild#446)

**Comment by gkhanna79** [comment]
We have CoreCLR repo generating nuget packages for Windows/OSX/Ubuntu.

**Comment by gkhanna79** [comment]
This work is complete for all platforms but OSX that is running into VSO build issues (which dont repro in CI or locally).

**Comment by gkhanna79** [comment]
OSX build has been fixed as well and all open builds are generating nuget packages that contain mscorlib.ni.dll.

Closing this issue now.
